### DISCUSSION TITLE
How to create a standalone prg executable of my code?

### DISCUSSION THREAD

**Discussion by sanmes** [jkotlinski/durexforth#491] [Q&A] [answered]
I would like to know how to create a standalone program from the words I have created and execute the word MAIN once loaded. 

I have tried to follow the instructions in the manual but I failed miserably, I find the info a bit hard to understand, I am new to forth.

I have also looked at #382 10 print turnkey demo  and I think that this is what I need in my code but I don't know how to adapt it to my needs: 

```
include turnkey
' 10print start !
9 device
save-prg 10print
8 device
```
I tried to add this to my code changing 10print to main but it didn't work...

```
include turnkey
' main start !
save-prg main
```
I get the error 62, file not found

Thanks

**Comment by sanmes** [comment, marked as answer]
I have found why I was getting the error

I was using the durex forth cartridge and I needed the disk to include turnkey !

 and I have an executable now working.

**Comment by Whammo** [comment]
Good job. How is your experience with durexForth?

**Reply by sanmes** [reply]
I haven't coded much yet but it is very interesting indeed. I am wondering how fast it can be. I heard that forth can be sometimes close to asm in speed. I have coded some words to make the core of a game, so I am printing chars and tiles to the screen in pure forth at the moment just to check how fast it can go. I am moving two tiles of 2x2 chars across the screen, deleting them before each move and it is taking more than a frame to execute so perhaps I will have to add assembler to bits that need the extra speed. And I am sure my code is not that good yet so it can be improved I have no doubt. 

For the C64 I usually code in assembler but I have also used in the past C + assembler and recently I have started to play with TRSE (Turbo Rascal Syntax Error) Pascal along routines in assembler and then I heard about Forth's small size and speed so I have start to code using Durex Forth.

BTW is there any way to change the address where code goes in memory? There is so much to learn. For instance I need to read more about 'here' and 'latest' as I didn't understand very well when I first came across the text mentioning them.

**Comment by Whammo** [comment]
Storing screen codes to screen memory rather than printing could be the fastest way to deal with that.

 ```here``` Is essentially the byte after the last byte of compiled code, the first byte of free contiguous memory.
```' here``` is approximately where it is stored
 ```latest``` Is the address of the last dictionary entry. So ```latest``` minus the length of the last dictionary entry, minus ```here```, is the amount of free contiguous memory.
 I have considered temporarily changing ```here``` to compile ASM code to say $c000 and then saving it to make it useable elsewhere, but I have not yet spent the time to see if there are complications.

**Reply by Whammo** [reply]
Yes, ```$c000 to here``` seems to have no problem. ```here``` is a system ```value```

**Reply by sanmes** [reply]
thanks for that! 

To print the char on the screen I use this word 

: pintachar ( char x y -- )
40 * + $0400 + c! ;

**Reply by Whammo** [reply]
If it were possible, do the math first, then do the store in the frame...?

**Comment by sanmes** [comment]
Good morning,
I have created this new painttile word, (the previous version used variables and  called 4 times the word paintchar,) it paints directly to the screen a 2x2 chars tile, no attributes, how can I improve it further, ie make it faster? without using assembler, thanks :

tscreen is a table of 16bit pointers to each row/line of the screen.
ttiles is a table with the tile info each tile is 4 bytes/chars

```
: painttile ( tile x y -- )
2* tscreen + @ + ( tile pscreen )
4 * ttiles + ( pscreen ptile )
2dup c@ swap c!  \ paint tile char 1 
2dup 1+ c@ swap 1+ c! \ paint tile char 2 
2dup 2+ c@ swap 40 + c! \ paint tile char 3 
3 + c@ swap 41 + c!  \ paint tile char 4 
;

```

**Reply by sanmes** [reply]
Perhaps I should try to use the return stack too?!... hm

**Reply by sanmes** [reply]
I was told by Master Mystic of the lemon64 forum to use @ and ! instead of c@ c! to fetch the chars and poke them into the screen area. It is faster now. 

```
: painttile ( tile x y -- )
2* tscreen + @ + ( tile pscreen )
swap (pscreen t )
4 * ttiles + ( pscreen ptile )
2dup @ swap !  \ paint tile char 1 and char2
2+ @ swap 40 + ! \ paint tile char 3 and char4
;
```

Are there any other possible improvements ?

**Reply by jkotlinski** [reply]
Hi and welcome!

I would guess `2* 2*` is faster than `4 *`. Otherwise, nothing obvious stands out.

Sadly, I suspect durexForth will not out-perform C or Pascal compilers, by size or speed. But maybe by fun?! ;-)

**Reply by Whammo** [reply]
If you do the paint first, having the stack already set up for it- literally two stores, then set up the stack for the next call to paint:
 if you're not just painting in a loop, it'll be faster.

**Reply by Whammo** [reply]
If you factor your code, you can eliminate swaps. 
 Pass the data in the order it is used.

**Comment by Whammo** [comment]
Untested.

```forth
: painttile ( tile x y -- )
2* tscreen + @ + >r
2* 2* ttiles + dup @ r@ !  \ oops fixed
2+ @ r> 40 + ! ;
```

**Reply by Whammo** [reply]
Three stack parameters are naturally problematic. One or two is what we shoot for.

**Reply by sanmes** [reply]
that's great thank you so much for that, I tried to use the return stack but I forgot completely about r@ so I was only using >r and r> and couldn't find a way for it to work. :)

**Comment by sanmes** [comment]
Thank you @jkotlinski , indeed I had the hope forth would be the fastest option, but on the other hand I am glad I have come to know a little about it as it is very interesting and very different from other languages I have used or use.

@Whammo What I understand is, I need to have all the calculations done, say the pointers and then just call the words to paint whatever is needed . This is what I do when not using double buffer in other languages, calculate everything that there is to calculate regading screen manipulation and then when the raster line reaches the low border I can paint everything to avoid flicker (when not using double buffer of course).

**Reply by Whammo** [reply]
I use forth to modify assembly code. Use one fast ASM routine, modify it's addresses and data with forth, then call it.

**Comment by sanmes** [comment]
Is the code produced by 2* the same as 1 lshift ? If so could we use 2 lshift instead of 2* 2* ?

**Comment by jkotlinski** [comment]
The result of ”2 lshift” and ”2* 2*” is the same. ”2* 2*” should be faster,
as it avoids putting a temporary 2 on the stack.

**Comment by sanmes** [comment]
![testforth](https://user-images.githubusercontent.com/37240460/201472789-853d9657-eb2b-496c-b8f3-52e256d6b6d4.jpg)

So this is the raster time taken to delete and paint two tiles, white and cyan. The red portion is the free raste time in a frame.

**Comment by sanmes** [comment]
I need to add sprite data and a charset to specific ram locations how is this done in Durex Forth? thx

I mean so that I have the data in one file without using loadb when the game runs. And how to prevent the game code to overwrite those areas?

**Reply by Whammo** [reply]
Here is how the author does sprites.
 
Calling the marker ```---editor---``` frees up a good deal of memory.
I don't think anyone has ever written a game with durexForth.
👍 
Perhaps ```save-forth``` ?

[durexforth](https://github.com/jkotlinski/durexforth)/[forth_src](https://github.com/jkotlinski/durexforth/tree/master/forth_src)/spritedemo.fs

**Reply by Whammo** [reply]
I guess it also depends where in memory you're to want them.

**Reply by sanmes** [reply]
Thank you @Whammo I am using the screen at $0400 so the data will need to be in that VIC bank too. I think using other banks will make things complicated further .

**Reply by Whammo** [reply]
How many sprites do you require? Are you using custom characters?

**Reply by sanmes** [reply]
A minimum of 15 sprites I think. Yes I'll be using a custom character set.

**Reply by Whammo** [reply]
Block 0 is pretty tight. 15 sprites and custom characters won't fit.

**Reply by Whammo** [reply]
Block 1 has the most room. No character ROM in 1 but you're using custom so okay.

**Reply by sanmes** [reply]
If I want to use bank 1 then how to do I setup durex forth to make code not everlap data etc? thx

**Reply by Whammo** [reply]
You get to manage memory yourself. All part of a bleeding edge programming environment, on a trailing edge system.
 Forth is kind of like assembly macros on steroids.

**Reply by sanmes** [reply]
So, how do I tell forth to use a certain memory and not other?

**Reply by Whammo** [reply]
durexForth compiles to ```here``` . You can compile your data by changing ```here``` to where you want it to go, or disk load it.

**Reply by sanmes** [reply]
Thanks a lot I will try that.

**Reply by sanmes** [reply]
Can you show me please a small example of using `here` in this context? I tried to change `here` to another address but didn't see any difference.

**Reply by sanmes** [reply]
As this will be just a small project I think I will use $e000 for the screen and the same bank for the chars and sprites. I will have just to copy over the char data and sprite data from where the data is stored by forth to those addresses. This is not optimal but for a small project will be enough and then I won-t have any problems with the dictionary overwriting data somewhere.

**Reply by sanmes** [reply]
I have started to code the game. I have the screen at $e000 and the charset at $e800 and the sprites from $eb00 onwards.
At the moment I have the code to print tiles on screen, read the level data from the buffer and paint the level on screen using the painttile word. I am also displaying a sprite. Not much yet but I am quite happy with the result. Check the video:
https://youtu.be/0sO1_cNU6No

**Reply by Whammo** [reply]
![here](https://user-images.githubusercontent.com/38296423/202308831-c10cb9aa-f336-41ff-93ac-a98f7eee7ae6.jpg)

**Reply by sanmes** [reply]
Oh I see, here is a value I was using ! when trying to change it. Thank you @Whammo!

**Comment by sanmes** [comment]
Here you can see what I have done up to now in my game written in Durex Forth: https://youtu.be/MVJj0l9JW5c

I have added the vertical movement and animation to baddies.

**Comment by jkotlinski** [comment]
Wow! That looks amazing! Great job!

**Reply by sanmes** [reply]
thanks I am focusing on the coding part, I will add more colour later

**Comment by sanmes** [comment]
I have added colour to the tiles: https://www.youtube.com/watch?v=PHorOC0ZlxU

One question, I am checking if a value is 1, 2 or 4. what I am doing is:

```
  dup dup
  0 > swap 3 < and swap 4 = or
  if
...
```
Is there a better way of doing it? If I use 'case' will it be faster ? thx

**Reply by Whammo** [reply]
```forth
7 and if
case
1 of ." one" endof
2 of ." two" endof
4 of ." four" endof
endcase
       then

```

**Comment by jkotlinski** [comment]
The standard way:

: 124? case
1 of -1 endof
2 of -1 endof
4 of -1 endof
0 swap endcase ;

Not sure its strictly better or just different…

**Comment by sanmes** [comment]
Thank you kindly @Whammo and @jkotlinski ! I will give your suggestions a try.

**Reply by Whammo** [reply]
Perhaps,

```forth
$f8 or if else
case
1 of ." one" endof
2 of ." two" endof
4 of ." four" endof
endcase
      then
```

**Reply by sanmes** [reply]
thank you

**Reply by Whammo** [reply]
It's all in what should be fastest, a hit or a miss.

**Comment by sanmes** [comment]
What is the best way to check the speed of a certain word, is there some kind of profiler?

**Comment by Whammo** [comment]
Jiffy clock? https://github.com/jkotlinski/durexforth/blob/master/forth_src/timer.fs

#335  or TOD?

**Comment by sanmes** [comment]
I have been optimizing words and I am happy with the result. I also have been taking my time to study CREATE and DOES>, not the easiest of topics but this is quite poweful.
I have done some more of the game, we can now control the player and he can collect objects:
https://youtu.be/1rPqNwln-VA

**Reply by Whammo** [reply]
This is a nice example of what can be done with ```create``` and ```does>```
 
 https://github.com/jkotlinski/durexforth/blob/master/forth_src/asm.fs
 
 It ties in knowledge of assembly so we know what it's supposed to do already.

**Reply by sanmes** [reply]
thank you

**Comment by jkotlinski** [comment]
Wow, that’s so cool!
That’s the first 8-bit Forth platformer I have seen! Maybe a world first?

**Comment by sanmes** [comment]
Update.
The game engine is almost 100% ready and it is possible to finish a level now.
I have lost some more speed though, I am hoping to squeeze our some speed when optimizing the new routines.
See the video here: https://www.youtube.com/watch?v=C9M990lkI0c

There is an issue though. I am coding using a text editor on the pc and then I paste the code in the Durex forth editor but now when I save the code to a d64 I get no error but when I try to load the file I get the error "too big" :|

**Comment by jkotlinski** [comment]
Right… the durexforth editor tries to load the entire file into RAM, but
there is only so much RAM available.
The workaround is to split the game into multiple, smaller files. Or use
another text editor altogether.

**Reply by sanmes** [reply]
I am using Visual Studio Code on the pc and pasting on vice, so I won't have problems if I continue to do it like this? I hope so :)

**Comment by jkotlinski** [comment]
How big is the file now, in kilobytes?

**Reply by sanmes** [reply]
11.0 KB (11,304 bytes)

**Comment by jkotlinski** [comment]
If you are using a PC for text editing, perhaps this technique can be
useful for you? It is a way to configure durexforth to load and evaluate a
given source file automatically at start up:

https://jkotlinski.github.io/durexforth/#_custom_start_up

Then, you would copy the source file to the .d64 image using a program such
as c1541.

**Reply by sanmes** [reply]
thanks that's very useful

**Comment by sanmes** [comment]
Update

https://www.youtube.com/watch?v=kVgsS6nDCFM

Now it will be the hardest part for me: design levels.

**Comment by sanmes** [comment]
update : https://youtu.be/_wzHEYgT--4
I am creating levels but I had some time to include the acid tiles animation and also to centre the play area and change the scoreboard.

**Comment by sanmes** [comment]
Which zero page addresses are free to use? I need to add a music player to the game but it needs some zp addresses to work. Thanks

**Comment by jkotlinski** [comment]
https://jkotlinski.github.io/durexforth/#_memory_map

**Reply by sanmes** [reply]
thanks!

**Comment by sanmes** [comment]
Just to confirm, as I am not using the editor can I use the memory from $a000 to $cfff for data/machine code?

**Comment by jkotlinski** [comment]
Yes, that is correct :-)

**Comment by sanmes** [comment]
I presume that it is not possible to bank out the kernal area $e000 - $ffff?! so I will have to use the irq vectores at $314, $315 instead..

**Comment by jkotlinski** [comment]
It is possible to temporarily swap out the kernal. Look at kernal-in,
kernal-out in gfx.fs

**Comment by sanmes** [comment]
I have a problem with turn-key again

The game is working fine when running from the interpreter, but when I do:
' run start !
save-prg game

the file game.prg loads fine and the game loads the files it needs but then it hangs.

I tried to use save-pack instead but I get the same result.

I am using interrupts and also loading stuff to $a000, $e800,$f000:
 s" mplayer" $a000 loadb drop
 s" chars" $e800 loadb drop
 s" tiles" tiles loadb drop
 s" sprites" $f000 loadb drop

Any clue to what might be happening here? Thanks

**Reply by Whammo** [reply]
Compile everything before overwriting the dictionary?

**Comment by sanmes** [comment]
Yes, I paste all the code in the interpreter, I think this way all is
compiled no?!

**Reply by Whammo** [reply]
If you wish to test your program after saving, you can compile a call to save-prg instead:

: build save-prg mydemo start @ execute ;
build
This will simulate the start-up sequence after saving the packed program.

**Comment by sanmes** [comment]
it worked fine when using:

: build save-prg run start @ execute ;
build

run is the word to launch the game

So what can be causing the problem when using to save the program file?:

' run start ! 
save-prg zezito


thanks

**Reply by Whammo** [reply]
I believe turnkey is included by default.

**Comment by sanmes** [comment]
I have found the problem!

I was using

' irq 

inside a word, instead of 

['] irq

**Reply by sanmes** [reply]
so this was what you @Whammo meant by your statement : Compile everything before overwriting the dictionary 

I am just a noob and somethings just go over my head still

**Comment by jkotlinski** [comment]
Nice find!

tis 24 okt. 2023 kl. 22:55 skrev S Mestre ***@***.***>:
